### PR TITLE
feat: SFT convergence run changes

### PR DESCRIPTION
**Pull request by yfw**
# What does this PR do ?

Several changes for SFT convergence run:

- Updated NLLLoss to take average loss over unmasked tokens (instead of sum)
- Updated sft.yaml to be consistent with default NeMo 2 llama3-8b recipe
- Added configurable optimizer (instead of hardcoded AdamW)
- Add a set_seed util for reproducible runs
- Update squad template to be consistent with NeMo 2 (add a space before answer)

# Changelog 
- Please update the [CHANGELOG.md](/CHANGELOG.md) under next version with high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation? Make sure to also update the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/index.html) which contains the tutorials

# Checklist when contributing
- [ ] TBD

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo_reinforcer/algorithms/sft.py [hunk 1]
```diff
@@ -434,6 +442,11 @@ def sft_train(
         logger.log_metrics(metrics, step + 1, prefix="train")
         logger.log_metrics(timing_metrics, step + 1, prefix="timing/train")
 
+        # To match NeMo2
```
Comment (parthchadha) [inline nemo_reinforcer/algorithms/sft.py]: Lets remove NeMo references?

File: nemo_reinforcer/algorithms/loss_functions.py [hunk 2]
```diff
@@ -160,6 +160,10 @@ def __call__(
 
         # Only compute loss on generated tokens (not input tokens)
         # by applying the token_loss_mask (shifted by 1 since we're predicting next tokens)
-        loss = -torch.sum(token_logprobs * mask)
+        num_unmasked_tokens = torch.sum(mask)
```
Comment (parthchadha) [inline nemo_reinforcer/algorithms/loss_functions.py:163]: I guess you would need to update NLL test?

File: examples/configs/sft_nemo_verify.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# SFT Algorithm Configuration
```
Comment (parthchadha) [inline examples/configs/sft_nemo_verify.yaml]: I would avoid publishing this config. We should have a single sft config.

Comment (SahilJain314) [reply examples/configs/sft_nemo_verify.yaml]: +1 please merge all of this into the main sft config. Also, please update the grpo config.

File: nemo_reinforcer/models/policy/hf_policy.py [hunk 4]
```diff
@@ -115,9 +115,17 @@ def do_fsdp(model):
         self._held_reference_model_params = None
         # register_fsdp_forward_method(self.model, "generate")
         if init_optimizer:
-            self.optimizer = torch.optim.AdamW(
-                self.model.parameters(), lr=self.cfg["learning_rate"]
-            )
+            if "optimizer" in self.cfg:
```
Comment (SahilJain314) [inline nemo_reinforcer/models/policy/hf_policy.py]: no code defaults! don't allow the user to not specify optimizer. Let's add it to all configs.

File: nemo_reinforcer/algorithms/sft.py [hunk 5]
```diff
@@ -331,6 +336,7 @@ def sft_train(
 
     policy.prepare_for_training()
 
+    consumed_samples = 0
```
Comment (ashors1) [inline nemo_reinforcer/algorithms/sft.py]: consumed_samples is tracked in the save_state now, so I think this is redundant

File: nemo_reinforcer/algorithms/sft.py (hunk 1 above)
Comment (yfw) [reply nemo_reinforcer/algorithms/sft.py]: Removed

File: examples/configs/sft_nemo_verify.yaml (hunk 3 above)
Comment (yfw) [reply examples/configs/sft_nemo_verify.yaml]: Sounds good, removed. For testing convergence against NeMo 2, I have some config overrides so that the settings match the NeMo 2 default finetuning recipe. Should we have some internal doc to keep track of that?

File: nemo_reinforcer/models/policy/hf_policy.py (hunk 4 above)
Comment (yfw) [reply nemo_reinforcer/models/policy/hf_policy.py]: Updated

File: nemo_reinforcer/algorithms/sft.py (hunk 5 above)
Comment (yfw) [reply nemo_reinforcer/algorithms/sft.py]: Removed

File: examples/configs/sft_nemo_verify.yaml (hunk 3 above)
Comment (yfw) [reply examples/configs/sft_nemo_verify.yaml]: Updated the default sft.yaml to reflect the convergence run

File: nemo_reinforcer/algorithms/loss_functions.py (hunk 2 above)
Comment (yfw) [reply nemo_reinforcer/algorithms/loss_functions.py:163]: Updated
